### PR TITLE
`Operator` defines `_flatten` and `_unflatten` methods

### PR DESCRIPTION
**Pull request by albi3ro**
See PR #4198  for the end goal. [sc-40554]

This PR adds `Operator._flatten` and `Operator._unflatten` private methods.

It then implements these methods for all operators that need custom implementations. 

A future PR will create an `Operator.__init_subclass` in order to register the operations as pytrees with jax.

Complications:

The `RandomLayers` keyword argument `ratio_imprim` was called `ratio_imprimitive` in the hyperparameters. This PR makes the name for that keyword argument consistent.

We have not yet registered tapes as pytrees.
The Hilbert-Schmidt template accepts a quantum tape as a argument.  This tape could potentially be a pytree with additional trainable information, but since tapes will not be registered as pytrees for another few months, we cannot include it as a leaf.

This shines light on a cyclical dependency issue.  Tapes should depend on operators, and operators should not depend on tapes.

### CODE REVIEW

File: doc/development/deprecations.rst [hunk 1]
```diff
@@ -6,6 +6,13 @@ Deprecations
 Pending deprecations
 --------------------
 
+* The `RandomLayers.compute_decomposition` keyword argument `ratio_imprivitive` has been changed to `ratio_imprim` to
+  match the call signature of the operation. 
+
+* The ``do_queue`` keyword argument in ``qml.operation.Operator`` is deprecated. This affects
+  all child classes, such as ``Operation``, ``Observable``, ``SymbolicOp`` and more. Instead of
+  setting ``do_queue=False``, use the ``qml.QueuingManager.stop_recording()`` context.
+
```
Comment (albi3ro) [inline doc/development/deprecations.rst:15]:
```suggestion
```

File: pennylane/templates/embeddings/angle.py [hunk 2]
```diff
@@ -20,6 +20,7 @@
 from pennylane.operation import Operation, AnyWires
 
 ROT = {"X": RX, "Y": RY, "Z": RZ}
+INV_ROT = {RX: "X", RY: "Y", RZ: "Z"}
```
Comment (albi3ro) [inline pennylane/templates/embeddings/angle.py:23]:
```suggestion
```

File: pennylane/templates/tensornetworks/ttn.py [hunk 3]
```diff
@@ -56,33 +56,21 @@ def compute_indices(wires, n_block_wires):
     n_wires = 2 ** (int(np.log2(len(wires) / n_block_wires))) * n_block_wires
     n_layers = int(np.log2(n_wires // n_block_wires)) + 1
 
-    layers = [
```
Comment (albi3ro) [inline pennylane/templates/tensornetworks/ttn.py:59]: Basically just switching lists to tuples.

But I couldn't help myself and had to fix this block of code.

File: doc/development/deprecations.rst [hunk 4]
```diff
@@ -6,6 +6,12 @@ Deprecations
 Pending deprecations
 --------------------
 
+* The `RandomLayers.compute_decomposition` keyword argument `ratio_imprivitive` has been changed to `ratio_imprim` to
```
Comment (trbromley) [inline doc/development/deprecations.rst:9]:
```suggestion
* The `RandomLayers.compute_decomposition` keyword argument `ratio_imprivitive` will be changed to `ratio_imprim` to
```

File: pennylane/templates/layers/random.py [hunk 5]
```diff
@@ -230,6 +232,13 @@ def compute_decomposition(
          CNOT(wires=['b', 'a']),
          RX(tensor(1.4000), wires=['a'])]
         """
+        if ratio_imprimitive:
+            warnings.warn(
+                "In RandomLayers.compute_decomposition, ratio_imprim has been changed to `ratio_imprimitive` to match the "
```
Comment (trbromley) [inline pennylane/templates/layers/random.py:237]:
```suggestion
                "In RandomLayers.compute_decomposition, ratio_imprim should be changed to `ratio_imprimitive` to match the "
```
Just to make it an explicit request to the user.

File: pennylane/operation.py [hunk 6]
```diff
@@ -1905,6 +1995,13 @@ class Tensor(Observable):
     tensor = True
     has_matrix = True
 
+    def _flatten(self):
```
Comment (albi3ro) [inline pennylane/operation.py:1998]: Really not sure why these lines aren't being marked as covered.

File: pennylane/templates/swapnetworks/ccl2.py [hunk 7]
```diff
@@ -84,6 +89,14 @@ class TwoLocalSwapNetwork(Operation):
     num_wires = AnyWires
     grad_method = None
 
+    @classmethod
+    def _unflatten(cls, data, metadata):
```
Comment (vincentmr) [inline pennylane/templates/swapnetworks/ccl2.py:93]: Is there a corresponding `flatten` method?

File: pennylane/templates/embeddings/displacement.py [hunk 8]
```diff
@@ -100,6 +100,12 @@ def circuit(feature_vector):
     num_wires = AnyWires
     grad_method = None
 
+    @classmethod
+    def _unflatten(cls, data, metadata):
```
Comment (vincentmr) [inline pennylane/templates/embeddings/displacement.py:104]: Do we need a corresponding `_flatten` method here?

File: pennylane/ops/qubit/arithmetic_ops.py [hunk 9]
```diff
@@ -359,6 +359,16 @@ class IntegerComparator(Operation):
 
     grad_method = None
 
+    def _flatten(self):
```
Comment (vincentmr) [inline pennylane/ops/qubit/arithmetic_ops.py:361]: Do we need `_unflatten` here?

File: pennylane/templates/embeddings/angle.py [hunk 10]
```diff
@@ -75,6 +75,13 @@ def circuit(feature_vector):
     num_wires = AnyWires
     grad_method = None
 
+    def _flatten(self):
```
Comment (vincentmr) [inline pennylane/templates/embeddings/angle.py:78]: Missing `_unflatten`?

File: pennylane/templates/embeddings/squeezing.py [hunk 11]
```diff
@@ -102,6 +102,12 @@ def circuit(feature_vector):
     num_wires = AnyWires
     grad_method = None
 
+    @classmethod
+    def _unflatten(cls, data, metadata) -> "SqueezingEmbedding":
```
Comment (vincentmr) [inline pennylane/templates/embeddings/squeezing.py:106]: Missing `_flatten`?

File: pennylane/ops/qubit/hamiltonian.py [hunk 12]
```diff
@@ -166,6 +166,14 @@ class Hamiltonian(Observable):
     batch_size = None
     ndim_params = None  # could be (0,) * len(coeffs), but it is not needed. Define at class-level
 
+    def _flatten(self):
+        # note that we are unable to restore grouping type or method without creating new properties
```
Comment (timmysilv) [inline pennylane/ops/qubit/hamiltonian.py:170]: is saving the `_grouping_indices` not enough? then in `_unflatten` you can say:
```
op = cls(data[0], data[1])
op._grouping_indices = metadata[0]
```

File: pennylane/templates/embeddings/displacement.py [hunk 13]
```diff
@@ -100,6 +100,12 @@ def circuit(feature_vector):
...
+        new_op = cls.__new__(cls)
```
Comment (timmysilv) [inline pennylane/templates/embeddings/displacement.py:105]: why was this necessary?

File: pennylane/templates/layers/random.py [hunk 14]
```diff
@@ -230,6 +232,13 @@ def compute_decomposition(
          CNOT(wires=['b', 'a']),
          RX(tensor(1.4000), wires=['a'])]
         """
+        if ratio_imprimitive:
```
Comment (timmysilv) [inline pennylane/templates/layers/random.py:235]: if this is set, and all previous values are defined positionally, won't they all be off-by-one and wrong?

**Review by timmysilv — COMMENTED** [review]
looks awesome! are we planning on integrating this with an actual (presumably Jax) pytree implementation soon to see that it works as expected? I'm assuming that won't be in this PR, but I'm curious about timeline/plans

File: pennylane/templates/embeddings/displacement.py (hunk 13 above)
Comment (albi3ro) [reply pennylane/templates/embeddings/displacement.py:105]: Easiest way to bypass the preprocessing done in `__init__`.  We already have the data, so we don't need to cast and stack it. `op.data` is not a valid `features`, so we have to do something else instead.

File: pennylane/templates/layers/random.py (hunk 14 above)
Comment (albi3ro) [reply pennylane/templates/layers/random.py:235]: If they defined it positionally, then its now `ratio_imprim` and no warning is raised.

```
qml.RandomLayers.compute_decomposition(weights,["a", "b"], 0.3,
                                    imprimitive=qml.CNOT, rotations=rotations, seed=42)
```
will raise no warnings.

Comment (timmysilv) [reply pennylane/templates/layers/random.py:235]: I guess if they start with `(weights,["a", "b"], ratio_imprimitive=0.3, ...` then python forces you to give names to all other kwargs so there should be no problem

File: tests/templates/test_subroutines/test_hilbert_schmidt.py [hunk 15]
```diff
@@ -19,6 +19,36 @@
...
+
+    assert hash(metadata)
+
+    new_op = type(op)._unflatten(*op._flatten())
+    assert qml.math.allclose(op.data, new_op.data)
+    assert op.hyperparameters["v_function"] == new_op.hyperparameters["v_function"]
+    assert op.hyperparameters["v_wires"] == new_op.hyperparameters["v_wires"]
+    for op1, op2 in zip(op.hyperparameters["u_tape"], new_op.hyperparameters["u_tape"]):
+        assert qml.equal(op1, op2)
+    assert new_op is not op
```
Comment (vincentmr) [inline tests/templates/test_subroutines/test_hilbert_schmidt.py:49]: We do not check for `assert qml.equal(op, new_op)`, is that because `qml.equal` does not support `HilbertSchmidt` objects, or just they should not be equal?

Comment (albi3ro) [reply tests/templates/test_subroutines/test_hilbert_schmidt.py:49]: `qml.equal` does not work with `HilbertSchmidt`.

Initially, I was going to add that change too in this PR, but decided there was already too much going on.

Basically, hyperparameters has `v_tape`, and tapes are only considered equal if they have equal identity.

**Comment by albi3ro** [comment]
> looks awesome! are we planning on integrating this with an actual (presumably Jax) pytree implementation soon to see that it works as expected? I'm assuming that won't be in this PR, but I'm curious about timeline/plans

Here's a branch where I have the next step: https://github.com/PennyLaneAI/pennylane/tree/register-operator-pytree

This next PR will go through sometime during this release cycle, but isn't blocking any other stories for the cycle.

**Review by vincentmr — APPROVED** [review]
Thanks for answering all my questions @albi3ro . The PR looks good to me, happy to approve.

File: pennylane/ops/qubit/hamiltonian.py (hunk 12 above)
Comment (albi3ro) [reply pennylane/ops/qubit/hamiltonian.py:170]: grouping indices is now metadata.  I needed to convert it all to tuples of tuples to get it to work, but should be fine now.

**Review by timmysilv — APPROVED** [review]
so cool!

File: pennylane/ops/op_math/exp.py [hunk 16]
```diff
@@ -167,6 +167,14 @@ class Exp(ScalarSymbolicOp, Operation):
     control_wires = Wires([])
     _name = "Exp"
 
+    def _flatten(self):
+        return (self.base, self.data[0]), (self.num_steps,)
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        print(data[1])
```
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:175]:
```suggestion
```
